src/vs/editor/contrib/inlayHints/utils.ts

```diff
@@ -0,0 +1,153 @@
+		rule.increaseRefCount();
+
+		return {
+			className: rule.className,
+			dispose: () => {
+				rule.decreaseRefCount();
+				this.garbageCollectionScheduler.schedule();
+			}
+		};
+	}
+
+	private getOrCreateRule(properties: CssProperties): RefCountedCssRule {
+		const key = this.computeUniqueKey(properties);
+		let existingRule = this.rules.get(key);
+		if (!existingRule) {
+			const counter = this.counter++;
+			existingRule = new RefCountedCssRule(key, `dyn-rule-${counter}`,
+				dom.isInShadowDOM(this.editor.getContainerDomNode())
+					? this.editor.getContainerDomNode()
+					: undefined,
+				properties
+			);
+			this.rules.set(key, existingRule);
+		}
+		return existingRule;
+	}
+
+	private computeUniqueKey(properties: CssProperties): string {
+		return JSON.stringify(properties);
+	}
+
+	private garbageCollect() {
+		for (const rule of this.rules.values()) {
+			if (!rule.hasReferences()) {
+				this.rules.delete(rule.key);
+				rule.dispose();
+			}
+		}
+	}
+}
+
+export interface ClassNameReference extends IDisposable {
+	className: string;
+}
+
+export interface CssProperties {
+	border?: string;
+	borderColor?: string | ThemeColor;
+	borderRadius?: string;
+	fontStyle?: string;
+	fontWeight?: string;
+	fontSize?: string;
+	fontFamily?: string;
+	textDecoration?: string;
+	color?: string | ThemeColor;
+	backgroundColor?: string | ThemeColor;
+	opacity?: string;
+	verticalAlign?: string;
+
+	margin?: string;
+	padding?: string;
+	width?: string;
+	height?: string;
+}
+
+class RefCountedCssRule {
+	private referenceCount: number = 0;
+	private styleElement: HTMLStyleElement;
+
+	constructor(
+		public readonly key: string,
+		public readonly className: string,
+		containerElement: HTMLElement | undefined,
+		public readonly properties: CssProperties,
+	) {
+		this.styleElement = dom.createStyleSheet(
+			containerElement
+		);
+
+		this.styleElement.textContent = this.getCssText(this.className, this.properties);
+	}
+
+	private getCssText(className: string, properties: CssProperties): string {
+		let str = `.${className} {`;
+		for (const prop in properties) {
+			const value = (properties as any)[prop] as string | ThemeColor;
+			let cssValue;
+			if (typeof value === 'object') {
+				cssValue = `var(${themeColorToCssVar(value)})`;
+			} else {
+				cssValue = value;
+			}
+
+			const cssPropName = camelToDashes(prop);
+			str += `\n\t${cssPropName}: ${cssValue};`;
+		}
+		str += `\n}`;
+		return str;
+	}
+
+	public dispose(): void {
+		this.styleElement.remove();
+	}
+
+	public increaseRefCount(): void {
+		this.referenceCount++;
+	}
+
+	public decreaseRefCount(): void {
+		this.referenceCount--;
+	}
+
+	public hasReferences(): boolean {
+		return this.referenceCount > 0;
+	}
+}
+
+function camelToDashes(str: string): string {
+	return str.replace(/(^[A-Z])/, ([first]) => first.toLowerCase())
+		.replace(/([A-Z])/g, ([letter]) => `-${letter.toLowerCase()}`);
+}
+
+function themeColorToCssVar(themeColor: ThemeColor): string {
```

Comment:
There is [`asCssVariableName`](https://github.com/microsoft/vscode/blob/d5db92d1f579fcc7fc1963a6baf16f39c0360833/src/vs/platform/theme/common/colorRegistry.ts#L34)